### PR TITLE
Implement own BERT score

### LINKED ISSUE #432: implement own BERT score
## 🚀 Feature

we shall come with our own BERT implementation as the bert_score require heavy Transformers dependency
follow-up for #424

### Motivation

keep the TM dependency light, and use other packages only as a reference

### Alternatives

use another package that has less dependencies and still use this one as reference

### LINKED ISSUE #472: BERTScore not working in DPP
## 🐛 Bug

`BERTScore` computation doesn't seem to work in DPP. If it's related, `bert_score` is currently returning scalar scores while several others such as `bleu_score` and `rouge_score` return tensor scores.

```python
import os
import torch.distributed as dist
import torch.multiprocessing as mp
from torchmetrics import BERTScore

def metric_ddp(rank, world_size):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12355"

    dist.init_process_group("gloo", rank=rank, world_size=world_size)

    metric = BERTScore(dist_sync_fn='mean')
    predictions = ["hello there", "general kenobi"]
    references = ["hello there", "master kenobi"]

    metric.update(predictions, references)
    print(f'rank {rank}: {metric.compute()}')

    dist.destroy_process_group()


if __name__ == "__main__":
    world_size = 2  # number of gpus to parallize over
    mp.spawn(metric_ddp, args=(world_size,), nprocs=world_size, join=True)

```

Running the above script results in the following traceback:
```
Traceback (most recent call last):
  File "/data/users/binhtang/scratch/bert_dpp.py", line 24, in <module>
    mp.spawn(metric_ddp, args=(world_size,), nprocs=world_size, join=True)
  File "/home/binhtang/local/miniconda3/envs/metrics/lib/python3.9/site-packages/torch/multiprocessing/spawn.py", line 230, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/home/binhtang/local/miniconda3/envs/metrics/lib/python3.9/site-packages/torch/multiprocessing/spawn.py", line 188, in start_processes
    while not context.join():
  File "/home/binhtang/local/miniconda3/envs/metrics/lib/python3.9/site-packages/torch/multiprocessing/spawn.py", line 150, in join
    raise ProcessRaisedException(msg, error_index, failed_process.pid)
torch.multiprocessing.spawn.ProcessRaisedException:

-- Process 1 terminated with the following error:
Traceback (most recent call last):
  File "/home/binhtang/local/miniconda3/envs/metrics/lib/python3.9/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
  File "/data/users/binhtang/scratch/bert_dpp.py", line 17, in metric_ddp
    print(f'rank {rank}: {metric.compute()}')
  File "/data/users/binhtang/metrics/torchmetrics/metric.py", line 362, in wrapped_func
    with self.sync_context(
  File "/home/binhtang/local/miniconda3/envs/metrics/lib/python3.9/contextlib.py", line 117, in __enter__
    return next(self.gen)
  File "/data/users/binhtang/metrics/torchmetrics/metric.py", line 333, in sync_context
    self.sync(
  File "/data/users/binhtang/metrics/torchmetrics/metric.py", line 285, in sync
    self._sync_dist(dist_sync_fn, process_group=process_group)
  File "/data/users/binhtang/metrics/torchmetrics/metric.py", line 240, in _sync_dist
    reduced = reduction_fn(output_dict[attr]) if reduction_fn is not None else output_dict[attr]
  File "/data/users/binhtang/metrics/torchmetrics/utilities/data.py", line 26, in dim_zero_cat
    x = [y.unsqueeze(0) if y.numel() == 1 and y.ndim == 0 else y for y in x]
  File "/data/users/binhtang/metrics/torchmetrics/utilities/data.py", line 26, in <listcomp>
    x = [y.unsqueeze(0) if y.numel() == 1 and y.ndim == 0 else y for y in x]
AttributeError: 'str' object has no attribute 'numel'
```

### Expected behavior

`{'precision': [1.0, 0.996446967124939], 'recall': [1.0, 0.996446967124939], 'f1': [1.0, 0.996446967124939], 'hashcode': 'roberta-large_L17_no-idf_version=0.3.10(hug_trans=4.9.2)_fast-tokenizer'}`

### Environment

- PyTorch Version (e.g., 1.0): 1.9.0
- OS (e.g., Linux): Linux
- How you installed PyTorch (`conda`, `pip`, source): `pip`
- Build command you used (if compiling from source): NA
- Python version: 3.9.6
- CUDA/cuDNN version: NA
- GPU models and configuration: NA
- Any other relevant information:

### Additional context

`pip install bert-score git+https://github.com/PyTorchLightning/metrics.git`

### PR DESCRIPTION
**Pull request by stancld**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?
- [x] Did you verify new and **existing tests pass** locally with your changes?

## What does this PR do?

This PR implements the `torchmetrics`' own BERTScore's and aims to remove the dependency on the **original** `bert-score` package (see https://github.com/Tiiiger/bert_score). Also, it aims to leave the `transformers` package as an optional one.

Fixes #432 resolves #472

**Current state:**: All important features should be implemented. Though it is a lot of code so I expect some changes are to be required after reviews :) 

## PR review

Anyone in the community is free to review the PR once the tests have passed. **Thank you for all the feedback in advance O:)**

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: metrics_examples/bert_score-own_model.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# Copyright The PyTorch Lightning team.
```
Comment (Borda) [inline metrics_examples/bert_score-own_model.py:1]: let's call the folder `tm_examples`

File: requirements/text.txt [hunk 2]
```diff
@@ -1,2 +1,3 @@
 nltk>=3.6
-bert-score==0.3.10
+transformers>=4.0
```
Comment (Borda) [inline requirements/text.txt:2]: can we have it as another optional? I guess the main point is to drop dependency on HF (not that I don't like nut lets try to be as light as possible - using PT only)

File: torchmetrics/functional/text/bert.py [hunk 3]
```diff
@@ -11,105 +11,667 @@
...
+            as an input and return the model's output represented by the single `torch.Tensor`.
+
+    Return:
+        A tuple of torch.Tensors containing the model's embeddings and matrix of tokens IDF. When `idf = False`,
+        each vector in model's embeddings is weighted by a factor of `1/sequence_length`.
+
+    Raises:
+        ValueError:
+    """
+    EMBEDDINGS_LIST: List[torch.Tensor] = []
```
Comment (Borda) [inline torchmetrics/functional/text/bert.py:305]: Capital vars are meant fro constant, here it is normal var

File: torchmetrics/functional/text/bert.py [hunk 4]
```diff
@@ -11,105 +11,667 @@
...
+        input_ids_idf /= input_ids_idf.sum(-1, keepdim=True)
+        IDF_SCALE_LIST.append(input_ids_idf)
+
+    EMBEDDINGS = torch.cat(EMBEDDINGS_LIST)
+    IDF_SCALE = torch.cat(IDF_SCALE_LIST)
+
+    return EMBEDDINGS, IDF_SCALE
+
+
+def _get_scaled_precision_or_recall(cos_sim: torch.Tensor, metric: str, idf_scale: torch.Tensor) -> torch.Tensor:
```
Comment (Borda) [inline torchmetrics/functional/text/bert.py:360]:
```suggestion
def _get_scaled_precision_or_recall(cos_sim: Tensor, metric: str, idf_scale: Tensor) -> Tensor:
```
lets simplify if and do `from torch import Tensor`

**Review by Borda — COMMENTED** [review]
I like it and see the huge effort behind it! thank you :rabbit:

**Comment by stancld** [comment]
> I like it and see the huge effort behind it! thank you 🐰

Thank you very much for the feedback. I'll have a look! :)

**Comment by stancld** [comment]
@Borda Do you think that it would be okay to mark this PR as ready for review to receive some initial feedback from the community? :)

File: tests/text/test_bertscore.py [hunk 5]
```diff
@@ -1,12 +1,21 @@
...
 
-from torchmetrics.functional import bert_score
+from torchmetrics.functional import bert_score as metrics_bert_score
 from torchmetrics.text import BERTScore
 from torchmetrics.utilities.imports import _BERTSCORE_AVAILABLE
 
+if _BERTSCORE_AVAILABLE:
+    from bert_score import score as original_bert_score
+
+os.environ["TOKENIZERS_PARALLELISM"] = "true"
```
Comment (Borda) [inline tests/text/test_bertscore.py:17]: rather simple bool, string may case problem with cases
```suggestion
os.environ["TOKENIZERS_PARALLELISM"] = 1
```

File: tm_examples/bert_score-own_model.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+_MODEL_DIM = 4
+_NHEAD = 2
+_MAX_LEN = 6
+
+
+class UserTokenizer:
+    def __init__(self):
+        self.cls_token = "<cls>"
+        self.sep_token = "<sep>"
+        self.pad_token = "<pad>"
```
Comment (Borda) [inline tm_examples/bert_score-own_model.py:37]:
```suggestion
    CLS_TOKEN = "<cls>"
    SEP_TOKEN = "<sep>"
    PAD_TOKEN = "<pad>"

    def __init__(self):
```
are they class constants or spec for each instance?

File: tm_examples/bert_score-own_model.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+from torchmetrics import BERTScore
+
+_NUM_LAYERS = 2
+_MODEL_DIM = 4
+_NHEAD = 2
+_MAX_LEN = 6
+
+
+class UserTokenizer:
```
Comment (Borda) [inline tm_examples/bert_score-own_model.py:33]: pls add docstring as it is an example user shall know what/why is this needed

File: torchmetrics/functional/text/bert.py [hunk 8]
```diff
@@ -11,105 +11,659 @@
...
+                Attention mask (`torch.Tensor`).
+            idf:
+                An indication of whether calculate token inverse document frequencies to weight the model embeddings.
+            tokens_idf:
+                Inverse document frequencies (these should be calculated on reference sentences).
+        """
+        self.text = {
+            k: v
+            for k, v in zip(["input_ids", "attention_mask"], _sort_data_according_length(input_ids, attention_mask))
+        }
```
Comment (Borda) [inline torchmetrics/functional/text/bert.py:241]:
```suggestion
        self.text = dict(zip(["input_ids", "attention_mask"], _sort_data_according_length(input_ids, attention_mask)))
```

File: torchmetrics/functional/text/bert.py [hunk 9]
```diff
@@ -11,105 +11,659 @@
...
+                An indication of whether a non-default user's own tokenizer is used.
+        """
+        self.text = preprocess_text_fn(text, tokenizer, max_length)
+        self.max_length = self.text["input_ids"].shape[1]
+        self.num_sentences = len(text)
+        self.idf = idf
+        if idf:
+            self.tokens_idf = tokens_idf if tokens_idf is not None else self._get_tokens_idf()
+        else:
+            self.tokens_idf = {}
```
Comment (Borda) [inline torchmetrics/functional/text/bert.py:177]:
```suggestion
        self.tokens_idf = {}
        if idf:
            self.tokens_idf = tokens_idf if tokens_idf is not None else self._get_tokens_idf()
```

File: torchmetrics/functional/text/bert.py [hunk 10]
```diff
@@ -11,105 +11,659 @@
...
+            for k, v in zip(["input_ids", "attention_mask"], _sort_data_according_length(input_ids, attention_mask))
+        }
+        self.text = _input_data_collator(self.text)
+        self.num_sentences = len(self.text["input_ids"])
+        self.max_length = self.text["input_ids"].shape[1]
+        self.idf = idf
+        if idf:
+            self.tokens_idf = tokens_idf if tokens_idf is not None else self._get_tokens_idf()
+        else:
+            self.tokens_idf = {}
```
Comment (Borda) [inline torchmetrics/functional/text/bert.py:249]:
```suggestion
        self.tokens_idf = {}
        if idf:
            self.tokens_idf = tokens_idf if tokens_idf is not None else self._get_tokens_idf()
```

File: torchmetrics/functional/text/bert.py [hunk 11]
```diff
@@ -11,105 +11,659 @@
...
+
+
+def _get_progress_bar(dataloader: DataLoader, verbose: bool = False) -> Union[DataLoader, tqdm.auto.tqdm]:
+    """Helper function returning either the dataloader itself when `verbose = False`, or it wraps the dataloader with
+    `tqdm.auto.tqdm`, when `verbose = True` to display a progress bar during the embbeddings calculation."""
+    if verbose:
+        return tqdm.auto.tqdm(dataloader)
+    else:
+        return dataloader
```
Comment (Borda) [inline torchmetrics/functional/text/bert.py:258]:
```suggestion
    return tqdm.auto.tqdm(dataloader) if verbose else dataloader
```

File: torchmetrics/functional/text/bert.py [hunk 12]
```diff
@@ -11,105 +11,659 @@
...
+            A user's own forward function used in a combination with `user_model`. This function must take `user_model`
+            and a python dictionary of containing `"input_ids"` and `"attention_mask"` represented by `torch.Tensor`
+            as an input and return the model's output represented by the single `torch.Tensor`.
+
+    Return:
+        A tuple of torch.Tensors containing the model's embeddings and matrix of tokens IDF. When `idf = False`,
+        each vector in model's embeddings is weighted by a factor of `1/sequence_length`.
+
+    Raises:
+        ValueError:
```
Comment (Borda) [inline torchmetrics/functional/text/bert.py:311]: in what ocasion?

**Review by Borda — APPROVED** [review]
Great work! Thank you :rabbit:

**Comment by Borda** [comment]
@stancld seems now taking way more time for all the latest configurations...

**Comment by stancld** [comment]
@Borda it seems some processes must hang out when running multiprocessing on MacOS outside the `if __name__ == "__main__"` clause. Looks like setting `join=False` (Perform a blocking join on all processes - [pytorch docs](https://pytorch.org/docs/stable/multiprocessing.html)) can, fortunately, help to solve this issue.

**Comment by Borda** [comment]
I see, lets address this issue in another PR :]
